Pull Request for Review of Project

aracelicaes opened the pull request:
This Project is an app built using Ruby on Rails meant to mimic the functionality of Eventbrite. This rails app was built to implement the creation of Private Events a Web app where you can create an event, "buy" your attendance for an event while being signed in by simply using your username.

TSE,
in order to be able to see the menu options, please ensure you:
"Create a User" and then "Login" with the user information. We recommend you create several users so you can create events with each user and review the functionality. Once you have several events created, your "Signed In" user can "BUY" the attendance for an event. This is how you'll be able to see the relationship between the event and user models by calling the attendance associations.
To access the event#show click on the event image.
To access the user#show click on the user image.

If you have any questions, please let us know,

Ara and Jessica

aracelicaes committed:
Initializing app

jessicafarias committed:
Adding User MVC via scaffold

jessicafarias committed:
Adding user_session_controller

Adding helpers into session_controller
Adding views and routes for session_controller

jessicafarias committed:
include helpers in controller

aracelicaes committed:
Generating Event Model & DB
Generating Event thru scaffold
Generating Event DB & migrated
Adding Associations in Event & User Models
Adding User Foreign Key to Event
Migrated DB
Addding creator alias to user in Event.

aracelicaes committed:
Linking Controllers
fixed Show in users_ctrlr.
Fixed #create in events_ctrlr.
Removed extra data from User views
Removed extra data from Events views
Assigning paths to Views.
Editing show.html for events & users.

aracelicaes committed:
Adding creator.name
Adding to Users show.html
Adding to Events show.html

jessicafarias committed:
Fixing views

Adding conditional to events list in user#show
Adding link to new event in user#index

jessicafarias committed:
Attendance MVC

Adding attendance references for user and event
database migrate
Adding associations for all models

jessicafarias committed:
Adding helper in events_helper

Trying to add a new attendance in show event

aracelicaes committed:
Added #Buy in Attendance
Adding buy route
Adding #buy in attendance
Adding #buy validations in ctrlr
Adding path in events#show
:

jessicafarias committed:
Editing Buy method for attendance

Change all events for each user#show

aracelicaes committed:
Add Validations on Events
Adding validations on events ctrlr for authentication
Editing Alert & Notice styles
Removing notice <p> from all views
Limiting #buy to current_user on views.

aracelicaes committed:
Display Attendance
Added queries User Ctrl
Added queries Event Ctrlr
Added attendees display on Event#show
Added atteded display on User#show
Modified User show.html
Modified Event show.html

jessicafarias committed:
Adding Scope for events

Adding upcoming events scope
Adding past events scope
Editing User show controller
Editing user#show view to display scopes

aracelicaes committed:
Modify Events
Adding Scope Queries Event#index
Adding Scope Views event_index.html

aracelicaes committed:
Import Bootstrap
Adding styles to all views
Adding NavBar
Importing Bootstrap
Separating logic from application.html view

jessicafarias committed:
Adding displaylist Helper

Update event#index view
Update event#show view

aracelicaes committed:
Adding styles import fontawesome

jessicafarias committed:
Adding displaylist_user helper to display users

jessicafarias committed:
Merge branch 'feature/events' of https://github.com/aracelicaes/private_events into feature/events

aracelicaes committed:
Rubocop & #login_helper
Created #login_helper
Inserted # in application.html
Fixed Rubocop Errors

aracelicaes committed:
Edit Readme

jessicafarias committed:
Fixing stylelint errors

aracelicaes committed:
Update README.md

Fix ReadME title

aracelicaes committed:
Adding Styles and Icons to Navbar and content

jessicafarias committed:
Adding rspec for user

Adding spec folder
Test validations for users

aracelicaes committed:
Add Validations for Events rspec

aracelicaes committed:
Changed Description titles for user & even specs

jessicafarias committed:
Fix user validation test

jessicafarias committed:
Testing associations for user model

aracelicaes committed:
Finished Event Validations & Associations rspec

aracelicaes committed:
Adding Attendance associations rspec

jessicafarias committed:
Adding capyvara tests for events and users

jessicafarias committed:
Merge changes

aracelicaes committed:
Adding System Specs
Adding Login spec
Adding Logout spec

aracelicaes committed:
Adding last system spec for login button

jessicafarias committed:
Fixing rubocop errors

jessicafarias committed:
Update Readme and session helper

Update current_user? helper
Update Readme.md

jessicafarias committed:
Adding test.yml

jessicafarias committed:
Update test.yml

jessicafarias committed:
Change ruby version

jessicafarias committed:
Change test.yml

jessicafarias committed:
Fix All linters

kobiyoyo commented on app/models/event.rb:
```diff
@@ -0,0 +1,13 @@
+  has_many :attendances
+  has_many :attendees, through: :attendances, source: :user
+
+  scope :upcoming_events, -> { where('date >= ?', Date.today) }
+  scope :past_events, -> { where('date < ?', Date.today) }
```
Lets use time instead of date for example `Time.zone.now`

aracelicaes replied:
Applied change.

kobiyoyo commented on app/models/user.rb:
```diff
@@ -0,0 +1,9 @@
+  validates :name, presence: true
+  validates :username, presence: true, length: { in: 4..12 }, uniqueness: true
+
+  has_many :events
```
- [ ] This should have a  `foreign_key: :creator_id` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

jessicafarias replied:
The foreign_key, as the name indicates, comes from an external/foreign element. 
In this case, we utilize the user_id as a foreign_key due to the fact that the user_id exists as part of the User. 
"create" is an alias or a way to RE-NAME that relationship existing between the user and the event. 
So, if we write Event.first.create we will obtain the user that created this event. 
If we add a column named creator_id in the events table, it will not be able to accomplish the essence of a foreign key, 
since it will belong to the same event table, which doesn't fit the logic nor the purpose behind using a foreign_key. 
Also, the Odin Project specifications do not explicitly entail the name that must be used as the foreign_key, and in any case, then creator_id should belong to the user and not the event. Please see reference:
https://www.w3schools.com/sql/sql_foreignkey.asp

kobiyoyo replied:
- [ ] Please check the screenshot I provided .

jessicafarias replied:
Please read this example, of RUBY GUIDES
** 3.5 Bi-directional Associations Ruby on Rails guides**
https://guides.rubyonrails.org/association_basics.html#bi-directional-associations.

kobiyoyo commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,82 @@
+
+  # GET /users/1
+  # GET /users/1.json
+  def show
+    user = User.find(params[:id]) # Encontramos usuario
+    @user_events = user.events unless user.nil? # Toda la lsita de eventos del usuario show
+    @user_attended = user.attended_events
+    @upcoming_events = @user_attended.upcoming_events
+    @past_events = @user_attended.past_events
+  end
```
For the `user` lets use `current_user` for example `current_user.events`

aracelicaes replied:
current_user is a built-in helper method we have created to help us identify the User that is in the current session (or currently logged in). If we were to use this, whenever we show a user, the issue would be that it would only show us the event pertaining to the currently signed in User. However, what we wish to achieve is to show every event with the information of the user that created SUCH events.  If we were to use current_user.events then it would only show the creator of the event if they're in the current session (currently signed in). Could you please help us understand why you wish us to implement this change?

kobiyoyo replied:
Here we want to get 
- [x] lists events created by user
- [x]  lists events previously attended by user using event model scope
- [x] lists upcoming events previously attended by user using event model scope
Thats why `current_user`  helper method is needed here

kobiyoyo commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,82 @@
+    end
+  end
+
+  # PATCH/PUT /users/1
+  # PATCH/PUT /users/1.json
+  # def update
+  #   respond_to do |format|
+  #     if @user.update(user_params)
+  #       format.html { redirect_to @user, notice: 'User was successfully updated.' }
+  #       format.json { render :show, status: :ok, location: @user }
+  #     else
+  #       format.html { render :edit }
+  #       format.json { render json: @user.errors, status: :unprocessable_entity }
+  #     end
+  #   end
+  # end
+
+  # DELETE /users/1
+  # DELETE /users/1.json
+  # def destroy
+  #   @user.destroy
+  #   respond_to do |format|
+  #     format.html { redirect_to users_url, notice: 'User was successfully destroyed.' }
+  #     format.json { head :no_content }
+  #   end
+  # end
```
Lets remove commented codes.

aracelicaes replied:
applied.

kobiyoyo reviewed (changes requested):
## Hi Team   ![](https://user-images.githubusercontent.com/38801423/62908122-703d4480-bd44-11e9-9115-42025ab7f341.png) 
#### Great Job so far, everything works fine, but there are a few more things required. 
### General

- [x] Instead of using buy  to attend an event, let's use something more descriptive like `attend`
- [x] The events table should have `creator_id` as the foreign key ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] The events table has column `creator_id` of type `integer`
* [x] An index for `creator_id` column is added
- [x] Lets use only Rspec and Capybara for testing
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

aracelicaes committed:
Fixing changes as per TSE
Removing all comments
Fixing date on event model to Time.now

aracelicaes commented:
>     * [ ]  Lets use only Rspec and Capybara for testing
Hi1 :smiley:  we only used Rspec and Capybara for this project. Is there any other place in the project that you saw we're using another gem? We tried checking but didn't find anything other than what Rspec and Capybara are and contain. We used the github repos (see attached imgs) as guides to build our tests, and no other resources, so we didn't think we were explicitly using another gem. Could you help us understand what you mean? thanks :smiley: 

![rspec](https://user-images.githubusercontent.com/35284745/93397182-b4035700-f83e-11ea-8370-8afe1c96d536.png)
![capybara](https://user-images.githubusercontent.com/35284745/93397185-b49bed80-f83e-11ea-82c3-447cd0aca589.png)

aracelicaes committed:
Fix errors as per TSE
Changed buy to attend in controller
Changed buy in views/events/show
Changed buy in routes
Changed buy in README

jessicafarias committed:
Fix stylelin errors

jessicafarias committed:
Change linter.yml

aracelicaes committed:
Upgrading Ruby Version to Ruby 2.7.1

aracelicaes committed:
Adding Stylelint as Dependency.

aracelicaes committed:
Update yarn.lock file

kobiyoyo commented:
> > ```
> > * [ ]  Lets use only Rspec and Capybara for testing
> > ```
> 
> Hi1  we only used Rspec and Capybara for this project. Is there any other place in the project that you saw we're using another gem? We tried checking but didn't find anything other than what Rspec and Capybara are and contain. We used the github repos (see attached imgs) as guides to build our tests, and no other resources, so we didn't think we were explicitly using another gem. Could you help us understand what you mean? thanks 
> 
> ![rspec](https://user-images.githubusercontent.com/35284745/93397182-b4035700-f83e-11ea-8370-8afe1c96d536.png)
> ![capybara](https://user-images.githubusercontent.com/35284745/93397185-b49bed80-f83e-11ea-82c3-447cd0aca589.png)

I can see a folder called `test` in your app , which appears when using `minitest`  framework.
